sympy/solvers/polysys.py

```diff
@@ -770,43 +762,69 @@ def factor_system_poly(polys: list[Poly]) -> list[list[Poly]]:
         raise TypeError("polys should be a list of Poly instances")
     if not polys:
         return [[]]
-    else:
-        base_domain = polys[0].domain
-        base_gens = polys[0].gens
-        if not all(poly.domain == base_domain and poly.gens == base_gens for poly in polys[1:]):
-            raise DomainError("All polynomials must have the same domain and generators")
 
-    eqs_factors = []
+    base_domain = polys[0].domain
+    base_gens = polys[0].gens
+    if not all(poly.domain == base_domain and poly.gens == base_gens for poly in polys[1:]):
+        raise DomainError("All polynomials must have the same domain and generators")
 
+    factor_sets = []
     for poly in polys:
         constant, factors_mult = poly.factor_list()
 
-        factors = [f for f, m in factors_mult]
-
         if constant.is_zero is True:
             continue
         elif constant.is_zero is False:
-            if not factors:
+            if not factors_mult:
                 return []
+            factor_sets.append({f for f, _ in factors_mult})
         else:
             constant = sqf_part(factor_terms(constant).as_coeff_Mul()[1])
             constp = Poly(constant, base_gens, domain=base_domain)
-            factors.insert(0, constp)
+            factors = {f for f, _ in factors_mult}
+            factors.add(constp)
+            factor_sets.append(factors)
 
-        eqs_factors.append(factors)
+    if not factor_sets:
+        return [[]]
 
-    return _cnf2dnf(eqs_factors)
+    return _factor_sets(factor_sets)
 
 
-def _cnf2dnf(eqs: list[list[Poly]]) -> list[list[Poly]]:
+def _factor_sets(factor_sets: list[set[Poly]]) -> list[list[Poly]]:
     """
-    Given a list of lists of Poly from the factorization of the equations in a
-    polynomial system, find the minimal set of factorised subsystems that is
-    equivalent to the original system. The result is a disjunctive normal form.
+    Helper that builds factor combinations.
     """
-    systems_set = {frozenset(sys) for sys in cartes(*eqs)}
-    systems = [s1 for s1 in systems_set if not any(s1 > s2 for s2 in systems_set)]
-    return _sort_systems(systems)
```

Comment:
I think we should keep the old function here as well and call it `_factor_sets_slow`. There could also be some test in the test suite that asserts that these two functions give equivalent results for randomly generated inputs or something.

If the call to `_sort_systems` is moved out to the caller then the function can be called with lists of lists of integers and then it can be tested with randomly generated lists of lists of integers.